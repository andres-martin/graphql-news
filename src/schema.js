const schema = `
    type Link {
        id: String!
        url: String
        description: String 
    }
`;

export default schema;
