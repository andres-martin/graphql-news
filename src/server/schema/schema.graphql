input AuthProvider {
  email: String!
  password: String!
}

type Comment {
  _id: ID!
  link: Link
  parent: Comment
  comments: [Comment]
  author(author: ID): User
  content: String!
  created_at: String!
}

type Link {
  _id: ID!
  author(author: ID): User
  comments: [Comment]
  created_at: String!
  description: String
  score: Int
  url: String!
}

input LinkFilter {
  top: Int
  urlContains: String
}

type User {
  _id: ID!
  about: String
  comments: [Comment]
  created_at: String!
  links: [Link]
  saved: [ID]
  username: String!
}

type SignInPayload {
  token: String!
  user: User!
}

type Query {
  allLinks(first: Int, skip: Int): [Link]
  link(_id: ID!): Link
  filterLinks(filter: LinkFilter): [Link]
  allUsers: [User]
  user(id: ID!): User
  allComments: [Comment]
}

type Mutation {
  createComment(link: ID!, parent: ID, content: String!): Comment
  createLink(url: String!): Link
  destroyLink(id: ID!): Link
  createUser(username: String!, provider: AuthProvider!): User
  signInUser(email: String!, password: String!): SignInPayload
  upvoteLink(_id: ID!): Link
  downvoteLink(_id: ID!): Link
}

type Subscription {
  vote(_id: ID!): Link
}

