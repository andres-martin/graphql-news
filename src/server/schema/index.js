import { GraphQLSchema } from 'graphql';
import QueryType from './query';
import MutationType from './mutation';
import SubscriptionType from './subscription';

const graphQLSchemaConfig = {
  query: QueryType,
  mutation: MutationType,
  subscription: SubscriptionType,
};

const schema = new GraphQLSchema(graphQLSchemaConfig);

export default schema;
