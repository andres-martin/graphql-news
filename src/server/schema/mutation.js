import { GraphQLID, GraphQLObjectType, GraphQLNonNull, GraphQLString } from 'graphql';
import { ObjectId } from 'mongodb';
import bcrypt from 'bcrypt';
import jsonwebtoken from 'jsonwebtoken';
import 'isomorphic-fetch';
import moment from 'moment';
import Mercury from '@postlight/mercury-parser';
import { AuthenticationError, UserInputError } from 'apollo-server';

import LinkType from './types/link';
import UserType from './types/user';
import CommentType from './types/comment';
import { AuthProvider, SignInPayload } from './types/auth';
import { LINK_VOTED, COMMENT_CREATED, pubsub } from '../constants';

const MutationType = new GraphQLObjectType({
  name: 'Mutation',
  fields: () => ({
    createComment: {
      type: CommentType,
      args: {
        link: { type: new GraphQLNonNull(GraphQLID) },
        parent: { type: GraphQLID },
        content: { type: new GraphQLNonNull(GraphQLString) },
      },
      resolve: async (_, data, { db: { Comments }, user }) => {
        if (!user) {
          throw new Error('You must be logged in to comment');
        }

        const { link, content, parent } = data;

        const comment = {
          author: ObjectId(user.id),
          comments: [],
          content,
          created_at: moment(Date.now()).format('{YYYY} MM-DDTHH:mm:ss SSS [Z] A'),
          link: ObjectId(link),
          parent: parent ? ObjectId(parent) : null,
        };

        const response = await Comments.insert(comment);

        const commentObj = Object.assign({ _id: response.insertedIds[0] }, comment);

        pubsub.publish(COMMENT_CREATED, { commentCreated: commentObj });
        console.log(pubsub.publish(COMMENT_CREATED, { commentCreated: commentObj }));

        return commentObj;
      },
    },
    createLink: {
      type: LinkType,
      args: {
        url: { type: new GraphQLNonNull(GraphQLString) },
      },
      resolve: async (_, { url }, { db: { Links }, user }) => {
        if (!user) {
          throw new Error('You must be logged in to submit links');
        }

        let linkTitle;

        const getTitle = async qs =>
          await Mercury.parse(qs).then(response => {
            const { title } = response;

            linkTitle = title;
          });

        await getTitle(url);

        const link = {
          author: user.id,
          created_at: moment(Date.now()).format('{YYYY} MM-DDTHH:mm:ss SSS [Z] A'),
          score: 0,
          comments: [],
          description: linkTitle,
          url,
        };

        const response = await Links.insert(link);

        return Object.assign({ _id: response.insertedIds[0] }, link);
      },
    },
    destroyLink: {
      type: LinkType,
      args: {
        id: { type: new GraphQLNonNull(GraphQLID) },
      },
      resolve: async (_, { id }, { db: { Links }, user }) => {
        if (!user) {
          throw new Error('You must be logged in to remove a link');
        }

        const link = await Links.findOne(ObjectId(id));

        if (!link) {
          throw new Error('Link does not exist');
        }

        const response = await Links.deleteOne({ _id: ObjectId(id) });

        if (response.deletedCount === 1) {
          return link;
        }

        throw new Error('Link could not be destroyed');
      },
    },
    createUser: {
      type: UserType,
      args: {
        username: { type: new GraphQLNonNull(GraphQLString) },
        provider: { type: new GraphQLNonNull(AuthProvider) },
      },
      resolve: async (_, { username, provider }, { db: { Users } }) => {
        const hash = await bcrypt.hash(provider.password, 10);

        const newUser = {
          username,
          email: provider.email,
          password: hash,
          created_at: moment(Date.now()).format('{YYYY} MM-DDTHH:mm:ss SSS [Z] A'),
        };
        const response = await Users.insert(newUser);

        return Object.assign({ _id: response.insertedIds[0] }, newUser);
      },
    },
    signInUser: {
      type: SignInPayload,
      args: {
        email: { type: new GraphQLNonNull(GraphQLString) },
        password: { type: new GraphQLNonNull(GraphQLString) },
      },
      resolve: async (_, { email, password }, { db: { Users } }) => {
        const user = await Users.findOne({ email });

        if (!user) {
          throw new UserInputError('No user exists with that email address. Please try again');
        }

        const valid = await bcrypt.compare(password, user.password);

        if (!valid) {
          throw new UserInputError('Incorrect password. Please try again.');
        }

        const token = jsonwebtoken.sign(
          {
            id: user._id,
            email: user.email,
          },
          process.env.JWT_SECRET,
          { expiresIn: '3w' }
        );

        return { token, user };
      },
    },
    upvoteLink: {
      type: LinkType,
      args: {
        // We can't vote on a link w/o its ID
        _id: { type: new GraphQLNonNull(GraphQLID) },
      },
      resolve: async (_, { _id }, { user, db: { Links } }) => {
        if (!user) {
          throw new AuthenticationError('You must be logged in to vote');
        }

        const link = await Links.findOneAndUpdate({ _id: ObjectId(_id) }, { $inc: { score: 1 } });

        const { score } = link.value;

        pubsub.publish(LINK_VOTED, { vote: { _id, score: score + 1 } });

        return Links.findOne(ObjectId(_id));
      },
    },
    downvoteLink: {
      type: LinkType,
      args: {
        _id: { type: new GraphQLNonNull(GraphQLID) },
      },
      resolve: async (_, { _id }, { user, db: { Links } }) => {
        if (!user) {
          throw new AuthenticationError('You must be logged in to vote');
        }

        const link = await Links.findOneAndUpdate({ _id: ObjectId(_id) }, { $inc: { score: -1 } });

        const { score } = link.value;

        pubsub.publish(LINK_VOTED, { vote: { _id, score: score - 1 } }); // eslint-disable-line prettier/prettier

        return Links.findOne(ObjectId(_id));
      },
    },
  }),
});

export default MutationType;
