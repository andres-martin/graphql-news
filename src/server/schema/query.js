import {
  GraphQLID,
  GraphQLObjectType,
  GraphQLNonNull,
  GraphQLInt,
  GraphQLList,
  GraphQLString,
  GraphQLInputObjectType,
  GraphQLBoolean,
} from 'graphql';
import { ObjectId } from 'mongodb';

import LinkType from './types/link';
import UserType from './types/user';
import CommentType from './types/comment';

const LinkFilter = new GraphQLInputObjectType({
  name: 'LinkFilter',
  fields: () => ({
    urlContains: { type: GraphQLString },
    top: { type: GraphQLBoolean },
    recent: { type: GraphQLBoolean },
    controversial: { type: GraphQLBoolean },
  }),
});

const QueryType = new GraphQLObjectType({
  name: 'Query',
  fields: () => ({
    link: {
      type: LinkType,
      args: {
        // We must know a link's ID in order to query for it
        _id: { type: new GraphQLNonNull(GraphQLID) },
      },
      resolve: async (_, { _id }, { db: { Links } }) => await Links.findOne(ObjectId(_id)),
    },
    links: {
      type: new GraphQLList(LinkType),
      args: {
        first: { type: GraphQLInt },
        skip: { type: GraphQLInt },
        filter: { type: LinkFilter },
      },
      resolve: async (_, { first, skip, filter }, { db: { Links } }) => {
        const { urlContains, top, recent, controversial } = filter;

        if (urlContains) {
          return await Links.find({ url: { $regex: `.*${urlContains}.*` } }).toArray();
        }

        let links = await Links.find({}).toArray();

        if (top) {
          const compareScore = (a, b) => {
            if (a.score > b.score) {
              return -1;
            }
            if (a.score < b.score) {
              return 1;
            }
            return 0;
          };

          links = links.sort(compareScore);
        }

        if (recent) {
          const compareDates = (a, b) => {
            if (a.created_at > b.created_at) {
              return -1;
            }
            if (a.created_at < b.created_at) {
              return 1;
            }
            return 0;
          };

          links = links.sort(compareDates);
        }

        if (controversial) {
          const compareComments = (a, b) => {
            if (a.comments.length > b.comments.length) {
              return -1;
            }
            if (a.comments.length < b.comments.length) {
              return 1;
            }
            return 0;
          };

          links = links.sort(compareComments);
        }

        if (first && (!skip || skip === 0)) {
          return links.slice(0, first);
        }
        if (skip && !first) {
          return links.filter((v, i) => i > skip);
        }
        if (skip && first) {
          return links.filter((v, i) => i > skip).slice(0, first);
        }

        return links;
      },
    },
    allLinks: {
      type: new GraphQLList(LinkType),
      resolve: async (_, data, { db: { Links } }) => await Links.find({}).toArray(),
    },
    allUsers: {
      type: new GraphQLList(UserType),
      resolve: async (_, data, { db: { Users } }) => await Users.find({}).toArray(),
    },
    user: {
      type: UserType,
      args: {
        // We must know a user's ID in order to query for him/her
        id: { type: new GraphQLNonNull(GraphQLID) },
      },
      resolve: async (_, { id }, { db: { Users } }) => await Users.findOne(ObjectId(id)),
    },
    allComments: {
      type: new GraphQLList(CommentType),
      resolve: async (_, data, { db: { Comments } }) => await Comments.find({}).toArray(),
    },
    commentsForLink: {
      type: new GraphQLList(CommentType),
      args: {
        // We must know a link's ID in order to query for its comments
        _id: { type: new GraphQLNonNull(GraphQLID) },
      },
      resolve: async (_, data, { db: { Comments } }) => {
        const comments = await Comments.find({}).toArray();
        // eslint-disable-next-line consistent-return,array-callback-return
        return comments.filter(i => {
          if (i.link) {
            return i.link.toString() === data._id && i.parent === null;
          }
        });
      },
    },
  }),
});

export default QueryType;
