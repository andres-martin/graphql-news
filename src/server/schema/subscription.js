import { GraphQLID, GraphQLObjectType, GraphQLNonNull } from 'graphql';

import LinkType from './types/link';
import CommentType from './types/comment';
import { LINK_VOTED, COMMENT_CREATED, pubsub } from '../constants';

const SubscriptionType = new GraphQLObjectType({
  name: 'Subscription',
  fields: () => ({
    vote: {
      type: LinkType,
      args: {
        _id: { type: new GraphQLNonNull(GraphQLID) },
      },
      subscribe: () => pubsub.asyncIterator(LINK_VOTED),
    },
    commentCreated: {
      type: CommentType,
      subscribe: () => pubsub.asyncIterator(COMMENT_CREATED),
    },
  }),
});

export default SubscriptionType;
