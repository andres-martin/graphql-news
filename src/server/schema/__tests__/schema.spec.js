import { graphql } from 'graphql';
import { addMockFunctionsToSchema } from 'graphql-tools';
import schema from '../index';

// eslint-disable-next-line
jasmine.DEFAULT_TIMEOUT_INTERVAL = 60000;

addMockFunctionsToSchema({
  schema,
  mocks: {
    Link: () => ({
      _id: 1,
      url: 'https://example.com',
      description: 'example',
      score: 6,
    }),
    User: () => ({
      _id: 2,
      username: 'example',
      about: 'lorem ipsum',
    }),
  },
});

describe('the graphql-news schema', async () => {
  test('can query for a link with a given _id', async () => {
    const query = `
      query taskForLink {
        link(_id: 1) {
          _id
          url
          description
          score
        }
      }  
    `;

    const res = await graphql(schema, query);

    expect(res.data).toMatchSnapshot();
  });

  test('can query for a list of links', async () => {
    const query = `
      query taskforAllLinks {
        allLinks {
          _id
        }
      }
    `;

    const res = await graphql(schema, query);

    expect(res.data).toMatchSnapshot();
  });

  test('can query for a user with a given _id', async () => {
    const query = `
      query taskForUser {
        user(_id: 2) {
          _id
          username
          about
        }
      }  
    `;

    const res = await graphql(schema, query);

    expect(res.data).toMatchSnapshot();
  });
});
