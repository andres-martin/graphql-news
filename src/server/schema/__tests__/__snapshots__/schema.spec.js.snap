// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the graphql-news schema can query for a link with a given _id 1`] = `
Object {
  "link": Object {
    "_id": "1",
    "description": "example",
    "score": 6,
    "url": "https://example.com",
  },
}
`;

exports[`the graphql-news schema can query for a list of links 1`] = `
Object {
  "allLinks": Array [
    Object {
      "_id": "1",
    },
    Object {
      "_id": "1",
    },
  ],
}
`;

exports[`the graphql-news schema can query for a user with a given _id 1`] = `
Object {
  "user": Object {
    "_id": "2",
    "about": "lorem ipsum",
    "username": "example",
  },
}
`;
